### PR TITLE
Search service proposal

### LINKED ISSUE #2180: Search service proposal
Write a proposal on how to extend Marquez searching capabilities.

### PR DESCRIPTION
**Pull request by pawel-big-lebowski**
Signed-off-by: Pawel Leszczynski <leszczynski.pawel@gmail.com>

### Problem

Search service proposal

Closes #2180

### Solution

Within PR.

> **Note:** All database schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

### Checklist

- [x] You've [signed-off](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#sign-your-work) your work
- [] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/MarquezProject/marquez/blob/main/CHANGELOG.md#unreleased) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned your `.sql` database schema migration according to [Flyway's naming convention](https://flywaydb.org/documentation/concepts/migrations#naming) (_if relevant_)
- [ ] You've included a [header](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#copyright--license) in any source code files (_if relevant_)

### CODE REVIEW

File: proposals/2180-search-service.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+size of such fields can be counted in megabytes. 
+
+[ElasticSearch](https://www.elastic.co) is widely used open-source search engine. It is beneficial to have its capabilities in Marquez because:
+ * It is designed to index and search JSON documents and OpenLineage events are JSONs by definition. 
+ * It is capable of searching by selected JSON fields or a whole OpenLineage event.
+ 
+### Pluggable Marquez Search Engine
+
+Some users find it extremely useful to use Marquez as simple UI for browsing OpenLineage events that is really easy
+easy to install. For those ElasticSearch can be an overkill. We still want to have an option to run Marquez without Search
```
Comment (wslulciuc) [inline proposals/2180-search-service.md:36]: Would the ElasticSearch and SimpleSearch implement an interface? If so, mind defining the contract of the interface?

File: proposals/2180-search-service.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+Service which will mean limited search capabilities but with simple installation steps.
+
+Similarly to database configuration, `yaml` configuration could contain extra section:
+```
+searchService:
```
Comment (wslulciuc) [inline proposals/2180-search-service.md:41]: I would recommend allowing other types of search engines (i.e. user implemented):

```yaml
search:
  engine: [ElasticSearch, ...]
  # engine specific properties
```

File: proposals/2180-search-service.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ ```
+  ResultType type; // enum DATASET or JOB
+  String name;
+  Instant updatedAt;
+  NamespaceName namespace;
+  NodeId nodeId;
+```
+ * In order to find datasets ES will query `dataset` index. When searching for jobs, `run` and `job` subtrees of `lineage` index will be queried. 
+ * A response from ES will contain runId for jobs, and (namespace, name) tuples for datasets. 
+ * Based on that, additional query to PostgreSQL will be sent to convert this into list of `SearchResult` entries.
```
Comment (wslulciuc) [inline proposals/2180-search-service.md:72]: What would be returned by querying postgres? Wouldn't the object be stored in ES and returned to the caller?

File: proposals/2180-search-service.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+Whenever lineage event is received in Marquez, it is sent to ElasticSearch indexes. Keep in mind that assuring uniqueness on
+Elastic side may be difficult, so in case of sending multiple same events Marquez will have a single copy of that (as long
+as it's a single run id), but SearchService will contain duplicate entries. 
+
+New `search` endpoint shall be implemented in Marquez and available only when `searchService` enabled.
+
+### Search Service in action
+
+ * Two indexes in ElasticSearch (ES) need to be created:
+   * `lineage`: to contain whole lineage events,
```
Comment (wslulciuc) [inline proposals/2180-search-service.md:57]: Maybe `lineage_events` as `lineage` might indicate that the caller can search the lineage graph?

File: proposals/2180-search-service.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+It's worth asking **Why PostgreSQL backend is not enough?**. 
+
+ * Because users want to search almost by anything (any table in our relational database) and being able to do SQL 
+`LIKE` on any column of any Marquez table could be a tedious task which could kill database at the end. 
+ * Within current Marquez architecture it is really important to have service up and running, so that incoming lineage
+events get saved in database. We don't want dummy search query to stop loading the events.
+ * We cannot create TEXT index on all the text fields. Creating index on jsonb columns could be dangerous as we know the 
+size of such fields can be counted in megabytes. 
+
+[ElasticSearch](https://www.elastic.co) is widely used open-source search engine. It is beneficial to have its capabilities in Marquez because:
```
Comment (wslulciuc) [inline proposals/2180-search-service.md:29]: Mind adding a digram and/or overview of deploying ElasticSearch along side Marquez and how meta would be indexed (i.e. index strategy)?

**Comment by mobuchowski** [comment]
I discussed this with Paweł offline, but I believe there are few things missing:

- we shouldn't index everything, but `whitelisted` facets
- we should know where we've found the keyword and somehow describe it in the search result

File: proposals/2180-search-service.md (hunk 3 above)
Comment (pawel-big-lebowski) [reply proposals/2180-search-service.md]: right, it's not necessary. `SearchResult` can be evaluated from ElasticSearch backend.

File: proposals/2180-search-service.md (hunk 5 above)
Comment (pawel-big-lebowski) [reply proposals/2180-search-service.md]: added one. let me know if you find it OK.

**Comment by pawel-big-lebowski** [comment]
> I discussed this with Paweł offline, but I believe there are few things missing:
> 
> * we shouldn't index everything, but `whitelisted` facets
Blacklisting `spark_plan` and `spark_unknown` definitely makes sense. I would be in favor of blacklisting just to allow searching on user's facets.
> * we should know where we've found the keyword and somehow describe it in the search result
As discussed, ElasticSeach highlight feature enriches document with `<em>` tag which is not super-clean and is tightly-coupled with ElasticSearch backend. I would prefer to start without it, we can add event preview on a list of search result to help user find out how the search lead to such results.

File: proposals/2180-search-service.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
+# Proposal: Search Service @ Marquez
+
+Author(s): @pawel-big-lebowski
+
+Created: 20022-10-20
```
Comment (mobuchowski) [inline proposals/2180-search-service.md:5]: One zero too much :)

File: proposals/2180-search-service.md (hunk 3 above)
Comment (wslulciuc) [reply proposals/2180-search-service.md]: I think as a follow up, we can discuss how we might want to revisit the `SearchResult` model and was to extend it.

File: proposals/2180-search-service.md (hunk 5 above)
Comment (wslulciuc) [reply proposals/2180-search-service.md]: ❤️

**Review by wslulciuc — APPROVED** [review]
A well thought-out and written proposal. Great job, @pawel-big-lebowski 💯 💯 🥇

**Pull request merged**